File: 3525-find-x-value-of-array-ii/3525-find-x-value-of-array-ii.java
class Solution {

    static class Node {
        int product;
        int[] cnt;

        Node(int k) {
            cnt = new int[k];
        }
    }

    int n, k;
    int[] nums;
    Node[] tree;

    public int[] resultArray(int[] nums, int k, int[][] queries) {
        this.n = nums.length;
        this.k = k;
        this.nums = nums;

        tree = new Node[4 * n];

        build(1, 0, n - 1);

        int[] ans = new int[queries.length];

        for (int q = 0; q < queries.length; q++) {

            int index = queries[q][0];
            int value = queries[q][1];
            int start = queries[q][2];
            int x = queries[q][3];

            // Persistent update
            nums[index] = value;
            update(1, 0, n - 1, index, value);

            // Query [start, n-1]
            Node res = query(1, 0, n - 1, start, n - 1);

            ans[q] = res.cnt[x];
        }

        return ans;
    }

    // --------------------------------------------------
    // BUILD
    // --------------------------------------------------

    void build(int node, int l, int r) {

        if (l == r) {
            tree[node] = makeLeaf(nums[l]);
            return;
        }

        int mid = l + (r - l) / 2;

        build(node * 2, l, mid);
        build(node * 2 + 1, mid + 1, r);

        tree[node] = merge(tree[node * 2], tree[node * 2 + 1]);
    }

    // --------------------------------------------------
    // LEAF
    // --------------------------------------------------

    Node makeLeaf(int value) {

        Node res = new Node(k);

        res.product = value % k;

        // Only one prefix: the element itself
        res.cnt[res.product] = 1;

        return res;
    }

    // --------------------------------------------------
    // MERGE
    // --------------------------------------------------

    Node merge(Node left, Node right) {

        Node res = new Node(k);

        // Product of the complete segment
        res.product =
            (int)((long)left.product * right.product % k);

        // Prefixes completely inside left
        for (int r = 0; r < k; r++) {
            res.cnt[r] += left.cnt[r];
        }

        // Prefixes which contain all of left
        // followed by a prefix of right
        for (int r = 0; r < k; r++) {

            int newRemainder =
                (int)((long)left.product * r % k);

            res.cnt[newRemainder] += right.cnt[r];
        }

        return res;
    }

    // --------------------------------------------------
    // UPDATE
    // --------------------------------------------------

    void update(int node, int l, int r,
                int index, int value) {

        if (l == r) {
            tree[node] = makeLeaf(value);
            return;
        }

        int mid = l + (r - l) / 2;

        if (index <= mid) {
            update(node * 2, l, mid, index, value);
        } else {
            update(node * 2 + 1, mid + 1, r, index, value);
        }

        tree[node] = merge(tree[node * 2],
                            tree[node * 2 + 1]);
    }

    // --------------------------------------------------
    // RANGE QUERY
    // --------------------------------------------------

    Node query(int node, int l, int r,
               int ql, int qr) {

        if (ql <= l && r <= qr) {
            return tree[node];
        }

        int mid = l + (r - l) / 2;

        if (qr <= mid) {
            return query(node * 2, l, mid, ql, qr);
        }

        if (ql > mid) {
            return query(node * 2 + 1, mid + 1, r, ql, qr);
        }

        Node left =
            query(node * 2, l, mid, ql, qr);

        Node right =
            query(node * 2 + 1, mid + 1, r, ql, qr);

        return merge(left, right);
    }
}

File: 3525-find-x-value-of-array-ii/3525-find-x-value-of-array-ii.cpp
class Solution {
public:

    struct Node {
        int prod;
        int cnt[5];

        Node() {
            prod = 1;
            for (int i = 0; i < 5; i++)
                cnt[i] = 0;
        }
    };

    int k;

    // --------------------------------------------------
    // Merge A + B
    // --------------------------------------------------
    Node mergeNode(const Node& A, const Node& B) {

        Node C;

        C.prod = (int)((long long)A.prod * B.prod % k);

        // Prefixes completely inside A
        for (int r = 0; r < k; r++) {
            C.cnt[r] = A.cnt[r];
        }

        // Prefixes = whole A + prefix of B
        for (int r = 0; r < k; r++) {

            int nr = (int)((long long)A.prod * r % k);

            C.cnt[nr] += B.cnt[r];
        }

        return C;
    }

    // --------------------------------------------------
    // Create leaf
    // --------------------------------------------------
    Node makeNode(int value) {

        Node res;

        res.prod = value % k;
        res.cnt[res.prod] = 1;

        return res;
    }

    // --------------------------------------------------
    // Solution
    // --------------------------------------------------
    vector<int> resultArray(
        vector<int>& nums,
        int k,
        vector<vector<int>>& queries
    ) {

        this->k = k;

        int n = nums.size();

        /*
            Iterative segment tree.

            Leaves:
                tree[n ... 2*n-1]

            Parents:
                tree[1 ... n-1]
        */

        vector<Node> tree(2 * n);

        // Build leaves
        for (int i = 0; i < n; i++) {
            tree[n + i] = makeNode(nums[i]);
        }

        // Build tree
        for (int i = n - 1; i >= 1; i--) {
            tree[i] = mergeNode(
                tree[i << 1],
                tree[i << 1 | 1]
            );
        }

        // --------------------------------------------------
        // Point update
        // --------------------------------------------------
        auto update = [&](int pos, int value) {

            pos += n;

            tree[pos] = makeNode(value);

            pos >>= 1;

            while (pos >= 1) {

                tree[pos] = mergeNode(
                    tree[pos << 1],
                    tree[pos << 1 | 1]
                );

                pos >>= 1;
            }
        };

        // --------------------------------------------------
        // Range query [l, r]
        // --------------------------------------------------
        auto query = [&](int l, int r) {

            Node leftResult;
            Node rightResult;

            bool hasLeft = false;
            bool hasRight = false;

            l += n;
            r += n + 1; // make r exclusive

            while (l < r) {

                if (l & 1) {

                    if (!hasLeft) {
                        leftResult = tree[l];
                        hasLeft = true;
                    }
                    else {
                        leftResult =
                            mergeNode(leftResult, tree[l]);
                    }

                    l++;
                }

                if (r & 1) {

                    --r;

                    if (!hasRight) {
                        rightResult = tree[r];
                        hasRight = true;
                    }
                    else {
                        rightResult =
                            mergeNode(tree[r], rightResult);
                    }
                }

                l >>= 1;
                r >>= 1;
            }

            if (!hasLeft)
                return rightResult;

            if (!hasRight)
                return leftResult;

            return mergeNode(leftResult, rightResult);
        };

        // --------------------------------------------------
        // Process queries
        // --------------------------------------------------

        vector<int> ans;
        ans.reserve(queries.size());

        for (auto& q : queries) {

            int index = q[0];
            int value = q[1];
            int start = q[2];
            int x = q[3];

            // Persistent update
            nums[index] = value;

            update(index, value);

            // Query [start, n-1]
            Node res = query(start, n - 1);

            ans.push_back(res.cnt[x]);
        }

        return ans;
    }
};